test/cmdLineTests/jvmtitests/src/com/ibm/jvmti/tests/nestMatesRedefinition/ClassGenerator.java

```diff
@@ -0,0 +1,370 @@
+/*******************************************************************************
+ * Copyright (c) 2018, 2018 IBM Corp. and others
+ *
+ * This program and the accompanying materials are made available under
+ * the terms of the Eclipse Public License 2.0 which accompanies this
+ * distribution and is available at https://www.eclipse.org/legal/epl-2.0/
+ * or the Apache License, Version 2.0 which accompanies this distribution and
+ * is available at https://www.apache.org/licenses/LICENSE-2.0.
+ *
+ * This Source Code may also be made available under the following
+ * Secondary Licenses when the conditions for such availability set
+ * forth in the Eclipse Public License, v. 2.0 are satisfied: GNU
+ * General Public License, version 2 with the GNU Classpath
+ * Exception [1] and GNU General Public License, version 2 with the
+ * OpenJDK Assembly Exception [2].
+ *
+ * [1] https://www.gnu.org/software/classpath/license.html
+ * [2] http://openjdk.java.net/legal/assembly-exception.html
+ *
+ * SPDX-License-Identifier: EPL-2.0 OR Apache-2.0 OR GPL-2.0 WITH Classpath-exception-2.0 OR LicenseRef-GPL-2.0 WITH Assembly-exception
+ *******************************************************************************/
+
+package com.ibm.jvmti.tests.nestMatesRedefinition;
+
+import org.objectweb.asm.*;
+import org.objectweb.asm.ClassWriter;
+import org.objectweb.asm.FieldVisitor;
+import org.objectweb.asm.MethodVisitor;
+import org.objectweb.asm.Opcodes;
+
+
+public class ClassGenerator implements Opcodes {
+
+	public static byte[] nestHostNoAttribute () throws Exception {
+		ClassWriter classWriter = new ClassWriter(ClassWriter.COMPUTE_MAXS | ClassWriter.COMPUTE_FRAMES);
+
+		classWriter.visit(V1_8, ACC_PUBLIC | ACC_SUPER, "NestHost", null, "java/lang/Object", null);
+		classWriter.visitSource("NestHost.java", null);		
+
+		MethodVisitor methodVisitor;
+		methodVisitor = classWriter.visitMethod(ACC_PUBLIC, "<init>", "()V", null, null);
+		methodVisitor.visitCode();
+		Label label0 = new Label();
+		methodVisitor.visitLabel(label0);
+		methodVisitor.visitLineNumber(2, label0);
+		methodVisitor.visitVarInsn(ALOAD, 0);
+		methodVisitor.visitMethodInsn(INVOKESPECIAL, "java/lang/Object", "<init>", "()V", false);
+		methodVisitor.visitInsn(RETURN);
+		methodVisitor.visitMaxs(1, 1);
+		methodVisitor.visitEnd();
+
+		classWriter.visitEnd();
+		return classWriter.toByteArray();
+	}
+	
+	public static byte[] nestHostNoAttributeForceRedefinition () throws Exception {
```

Comment:
It looks like the same basic class with a few modifications repeated over and over.  It could be written as:
```

public enum RedefinitionPhase { ORIGIINAL, REDEFINE };

public class ClassGenerator implements Opcodes {

	public static byte[] generateClass(String name, RedefinitionPhase phase, Attribute[] attributes) {
		ClassWriter classWriter = new ClassWriter(ClassWriter.COMPUTE_MAXS | ClassWriter.COMPUTE_FRAMES);

		classWriter.visit(V1_8, ACC_PUBLIC | ACC_SUPER, name, null, "java/lang/Object", null);
		switch(phase) {
		case ORIGINAL:
			/* Original has the source attribute */
			classWriter.visitSource(name + ".java", null);	
			break;
		case REDEFINE:
			/* REDEFINE doesn't have the source attribute to 
			 * ensure the VM can't skip the redefine 
			 */
			break;
		default:
			throw IllegalArgumentException("phase needs to be set correctly");
		}
		
		/* Generate base constructor which just calls super.<init>() */
		MethodVisitor methodVisitor = classWriter.visitMethod(ACC_PUBLIC, "<init>", "()V", null, null);
		methodVisitor.visitCode();
		methodVisitor.visitVarInsn(ALOAD, 0);
		methodVisitor.visitMethodInsn(INVOKESPECIAL, "java/lang/Object", "<init>", "()V", false);
		methodVisitor.visitInsn(RETURN);
		methodVisitor.visitMaxs(1, 1);
		methodVisitor.visitEnd();

		/* Generate passed in attributes if there are any */
		for (Attribute attr : attributes {
			classWriter.visitAttribute(attr);
		}

		classWriter.visitEnd();
		return classWriter.toByteArray();

	}

	/* What is the purpose of these two classes?  Why is the host with no nest being redefined */
	public static byte[] nestHostNoAttribute () throws Exception {
		return generateClass("NestHost", Phase.ORIGINAL, null);
	}
	
	public static byte[] nestHostNoAttributeForceRedefinition () throws Exception {
		return generateClass("NestHost", Phase.REDEFINE, null);
	}


	public static byte[] nestHostWithNestMembersAttribute () throws Exception {
		String[] nestMembers = {"A", "B", "C"};
		NestMembersAttribute attr = new NestMembersAttribute(nestMembers);

		return generateClass("NestHost", Phase.ORIGINAL, new Attribute[]{ attr });

	}
	
	public static byte[] nestHostWithNestMembersAttributeForceRedefinition () throws Exception {
		String[] nestMembers = {"A", "B", "C"};
		NestMembersAttribute attr = new NestMembersAttribute(nestMembers);

		return generateClass("NestHost", Phase.REDEFINE, new Attribute[]{ attr });
	}
...
```

Or something similar.  This makes it clear that its a small set of modifications to the base class and makes for less code to maintain.

Reply:
Will do; thanks.